File: 34.cpp
vector<vector<int>> FindPath(TreeNode *root, int expectNumber){
    vector<vector<int>> ans;
    if(root == nullptr)
        return ans;
    int path[10000];
    addPath(root, path, 0, ans, expectNumber);
    return ans;
}

int addPath(TreeNode *root, int *path, int pathNum,
                            vector<vector<int>> &ans, int expectNumber){
    if(root -> left == nullptr && root -> right == nullptr){
        path[pathNum++] = root -> val;
        int sum = 0;
        for(int i = 0; i < pathNum; ++i){
            sum += path[i];
        }
        if(sum == expectNumber){
            vector<int> currentPath;
            for(int i = 0; i < pathNum; ++i){
                currentPath.push_back(path[i]);
            }
            ans.push_back(currentPath);
        }
        return 0;
    }
    path[pathNum++] = root -> val;
    if(root -> left != nullptr){
        addPath(root -> left, path, pathNum, ans, expectNumber);
    }
    if(root -> right != nullptr){
        addPath(root -> right, path, pathNum, ans, expectNumber);
    }
    return 0;
}

File: 34.java
import java.util.ArrayList;
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;
 
    public TreeNode(int val) {
        this.val = val;
 
    }
 
}
*/
public class Solution {
    private ArrayList<ArrayList<Integer>> ans;
    private ArrayList<Integer> path;
     
    public ArrayList<ArrayList<Integer>> FindPath(TreeNode root, int target) {
        ans = new ArrayList<ArrayList<Integer>>();
        if (root == null || target <= 0)
            return ans;
        path = new ArrayList<Integer>();
        addPath(root, target, 0);
        return ans;
    }
    private void addPath(TreeNode root, int target, int sum) {
        if (root == null)
            return;
        path.add(root.val);
        if (root.left == null && root.right == null &&
            sum + root.val == target) {
            ArrayList<Integer> temp = new ArrayList<Integer>(path);
            ans.add(temp);
        }
        addPath(root.left, target, sum + root.val);
        addPath(root.right, target, sum + root.val);
        path.remove(path.size() - 1);
    }
}
